### PR TITLE
Changes to the UI: Mandatory wallet setup, bottom navigation bar and more

### PR DESCRIPTION
**Pull request by Tyratox**
This PR changes parts of the UI quite a bit. The most significant changes are listed below:

- Changed routing: Start the user on the screen to save a newly generated seed along with a button to use it. Provide two buttons: 1) to accept this seed and start using the application, 2) to navigate to a second screen where a previously saved seed can be set. This setup makes setting a seed mandatory and makes the often used case of using a new seed easier. Restoring a wallet using a known seed only happens once if a user has used the application before and wants to set it up on a new device or after re-installing it.
- Make the lao qr code scanning screen full screen
- Remove conformation screen when scanning a lao qr code when connecting. (Might be problematic for verifying authenticity but it is also questionable how reading a lao id would help for that. -> **Needs discussion**)
- Merge connection & launch screen. Most people only ever use the connect screen. The launch screen can now be accessed from the connection screen with an additional button press.
- Provide a button to change what camera is used on phones for scanning (flip camera)
- Move wallet buttons to an action sheet (https://developer.apple.com/design/human-interface-guidelines/ios/views/action-sheets/). To support android, this unfortunately required a new dependency (https://github.com/expo/react-native-action-sheet). But at least it is provided by expo:)
- Replaced top navigation bar with bottom navigation bar
- Fixed typing for navigation at the cost of defining the navigation structure statically in the core. (Not in the individual features as it would result in inter-feature dependencies)

Due to these changes the full UI is no longer consistent within itself (i.e. the Lao list still looks the same) but I plan to change the remaining screens as well. Just trying to split the changes into multiple PRs in a way that somehow makes sense.

Screenshots: (The color scheme as well as the icons are not fixed yet. In some places such as the button for the launch screen, the icons also need to be accompanied by a label.)

# Initial Screen
Screenshot 2022-05-22 at 18 13 11

# Restore seed
Screenshot 2022-05-22 at 18 13 15

# Lao List
Screenshot 2022-05-22 at 18 13 25

# Connect screen
Screenshot 2022-05-22 at 18 13 46

# Wallet Home
Screenshot 2022-05-22 at 18 14 01

# Wallet Action Sheet
Screenshot 2022-05-22 at 18 14 07

### CODE REVIEW

File: fe1-web/src/features/wallet/screens/WalletCreateSeed.tsx [hunk 1]
```diff
@@ -1,57 +1,147 @@
...
+            {STRINGS.wallet_welcome_heading}
+          </Text>
+          <Text style={[Typography.paragraph, Typography.negative]}>
+            {STRINGS.wallet_welcome_text_first_time}
+          </Text>
+          <Text style={Typography.paragraph}>
+            <Text style={Typography.negative}>
+              {STRINGS.wallet_welcome_text_wallet_explanation_1}
+            </Text>
+            <Text> </Text>
```
Comment (Xelowak) [inline fe1-web/src/features/wallet/screens/WalletCreateSeed.tsx:87]: Are those empty `Text`s useful?

**Review by Xelowak — COMMENTED** [review]
Phew, this was a long PR! Really nice work! Thanks a lot for working on UI/UX, as the app definitely needed this. 

I saw that you worked on the navigation part too (especially on params) and that's really great too! Did it finally remove [this annoying warning](https://github.com/dedis/popstellar/issues/787) when launching a LAO?
I think it could be nice to add some documentation about this, so that new students that want to add a screen can easily do it and understand how it works.

Do you think creating all these icon types is better than having one generic icon component that just takes a name, a size and a color as inputs? I feel like it would save a lot of code/components since the only thing that changes between them is the name of the icon. What could be done is having a "library" of names that we use directly in the generic component, so that we're sure to use the correct ones.

I appreciate the fact that you made everything easier to keep consistent by adding `border.ts`, `spacing.ts`, ... So thanks for that!

I know it's not the purpose of this PR, but since you changed a lot how `Text`'s style is assigned, it could be nice to add all UI text in `strings.ts`, so that we don't have strings put magically in the UI anymore. It can naturally be done in another PR though. I noted a few ones in my comments, but there are more of them.

I know that passing arrays as styles will apply the last one in the list, but is there a specifiy reason why you use this a lot? (I'm not familiar with this, so consider it as a question rather than a remark).

As I said on Slack, there seems to be an issue when you don't have a camera or you deny permissions to use it. It just displays a blank page and the navigation bar disappears.

It was definitely a significant size PR, so don't expect me to have caught everything :sweat_smile:, but overall it's nice!

File: fe1-web/src/core/components/Button.tsx [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    backgroundColor: Colors.accent,
+    borderWidth: Border.width,
+    borderRadius: Border.radius,
+  } as ViewStyle,
+  negative: {
+    borderColor: Colors.contrast,
+  } as ViewStyle,
+});
+
+const Button = (props: IPropTypes) => {
```
[the remark this answers is not in the document]
Comment (Tyratox) [reply fe1-web/src/core/components/Button.tsx:26]: Hmm that would be an option but then people looking for a button will just probably just use the first thing the IDE will suggest when writing Button (which will then be the react-native button).

But since the result will look different and it can be caught in a code review I think your suggestion makes sense!

File: fe1-web/src/features/home/screens/ConnectConfirm.tsx [hunk 3]
```diff
@@ -98,13 +101,19 @@ const ConnectConfirm = () => {
           onChangeText={(input: string) => setLaoId(input)}
           defaultValue={laoId}
         />
+
+        <Button onPress={onButtonConfirm}>
+          <Text style={[Typography.base, Typography.centered, Typography.negative]}>
+            {STRINGS.general_button_confirm}
+          </Text>
+        </Button>
```
[the remark this answers is not in the document]
Comment (Tyratox) [reply fe1-web/src/features/home/screens/ConnectConfirm.tsx:109]: Yes that is true :) 👍

File: fe1-web/src/features/home/screens/Launch.tsx [hunk 4]
```diff
@@ -57,32 +69,49 @@ const Launch = () => {
...
+
+          <Button onPress={onTestClearStorage}>
+            <Text style={[Typography.base, Typography.centered, Typography.negative]}>
+              [TEST] Clear (persistent) storage
+            </Text>
+          </Button>
+
+          <Button onPress={navigation.goBack}>
+            <Text style={[Typography.base, Typography.centered, Typography.negative]}>Cancel</Text>
+          </Button>
```
[the remark this answers is not in the document]
Comment (Tyratox) [reply fe1-web/src/features/home/screens/Launch.tsx:111]: I agree but did not dare to touch it haha :) Do we still need the two test buttons? Clearing storage can also just be done in the browser devtools and I have never tried `LocalMockServer.ts`, does it work and should it be kept?

File: fe1-web/src/features/wallet/screens/WalletCreateSeed.tsx (hunk 1 above)
Comment (Tyratox) [reply fe1-web/src/features/wallet/screens/WalletCreateSeed.tsx:87]: Yes, they add spaces :)

**Comment by Tyratox** [comment]
Yes it seems as if #787 was actually fixed :)

I am not sure what to do with the icons but I totally agree that is is far from ideal. The problem is that a given icon set usually does not all desired icons and thus some flexibility is required. I will try something but it will require some workarounds because of typescript 😅

Yes, I tried to put as many strings in `strings.ts` as I found but I kept missing some 😅 (And for testing strings I thought it is best to leave them since they should never be translated).

Regarding the style array syntax: In react native this combines the styles and applies all of them similar to using css classes. This way it is not necessary to create styles for all possible combinations but you can mix them on the fly. And as you said, in case of two styles defining the same property, the last array element takes precedence. (https://reactnative.dev/docs/style)

And I will look into the issue with the connect page, this should of course not happen :)

File: fe1-web/src/core/components/Button.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (Xelowak) [reply fe1-web/src/core/components/Button.tsx]: Yeah I get it, but then it can be confusing since 2 components will have the same name, and it will depend on where it has been imported, so renaming it is nice, thanks!

File: fe1-web/src/features/wallet/screens/WalletCreateSeed.tsx (hunk 1 above)
Comment (Xelowak) [reply fe1-web/src/features/wallet/screens/WalletCreateSeed.tsx:87]: Isn't there a more straighforward solution? I'm thinking about `{'\n'}` like in [there](https://infinitbility.com/how-to-use-br-tag-in-react-native).

File: fe1-web/src/features/evoting/components/ElectionNotStarted.tsx [hunk 5]
```diff
@@ -50,7 +50,13 @@ const ElectionNotStarted = ({ election, questions, isOrganizer }: IPropTypes) =>
         )}
         renderItem={({ item }) => <Text style={styles.textOptions}>{`\u2022 ${item}`}</Text>}
       />
-      {isOrganizer && <WideButtonView title="Open election" onPress={onOpenElection} />}
+      {isOrganizer && (
+        <Button onPress={onOpenElection}>
+          <Text style={[Typography.base, Typography.centered, Typography.negative]}>
+            Open election
```
[the remark this answers is not in the document]
Comment (Xelowak) [reply fe1-web/src/features/evoting/components/ElectionNotStarted.tsx]: No worries!

File: fe1-web/src/core/components/PoPButton.tsx [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    backgroundColor: Colors.accent,
+    borderWidth: Border.width,
+    borderRadius: Border.radius,
+  } as ViewStyle,
+  negative: {
+    borderColor: Colors.contrast,
+  } as ViewStyle,
+});
+
+const PoPButton = (props: IPropTypes) => {
```
Comment (Xelowak) [inline fe1-web/src/core/components/PoPButton.tsx:25]: Is this button without text used anywhere (I honestly didn't check)? Otherwise we can simply keep the `PoPTextButton` and remove this one.

**Review by Xelowak — COMMENTED** [review]
Thanks a lot for the changes!

Nice, then feel free to close #787 as soon as your PR is merged.

You're right, test strings don't need to be put in strings.ts since they are temporary.

Okay, thanks a lot for the explanation, it's clearer to me why you're using this.

File: fe1-web/src/core/components/Icon.tsx [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+import { ExtendType } from 'core/types';
+
+type IoniconNames = keyof typeof Ionicons['glyphMap'];
+
+type Ionicon = {
+  iconName: IoniconNames;
+  IconFamily: typeof Ionicons;
+};
+
+const iconNameMap = {
```
[the remark this answers is not in the document]
Comment (Tyratox) [reply fe1-web/src/core/components/Icon.tsx:15]: I guess it depends :) It is definitely complicated I agree. The reason was to still enable the IDE to check the "name" prop we are passing.

`expo/vector-icons` exports multiple icon families, `Ionicons` is just one of them and certain icons are not present in it. (Not important for this PR but for the next one since there we also use a few material ui icons)

Now my goal was to have a uniform interface no matter what icon family is used (implementation detail). This was possible with each Icon being a separate component and is now possible with this map. The problem is that the Icon component somehow needs to know which icon family the icon is from, i.e. `Ionicons` or `MaterialIcons`. Previously this was just defined in each file and now there is a map from "our name" to a tuple of icon family and the name this icon has inside this icon family. This way the Icon component can look up this map and use the correct combination of icon family and name. We could omit the `IonIcon` type but I thought it would help preventing mistakes since now the linter will tell you if there is an invalid combination of icon family and icon name. (`IoniconNames` is a union type of all valid icon names).

And because we have this map from our names to the tuples, the linter can also check whether a valid name was passed to the Icon component (Whose map entry points to a valid tuple).

So basically a lot of work so that the linter can check everything. But if you think this is not necessary and people using the Icon component will know what they are doing, we can remove all of this and have a simpler file.

In any case I agree that I should add comments to this file is it is definitely not trivial 😅

[the remark this answers is not in the document]
Comment (Xelowak) [reply fe1-web/src/core/components/Icon.tsx:38]: Mhmm ok, I understand why you did this now. Maybe another solution would be to create different icon components for each family? Imagine having a `Ionicon` component and a `MaterialIcon` component, so that you can omit the family and create 2 separated collections of names? This might not scale well for a lot of Icon families, but as I understand we're using two of them at the moment so it shouldn't be an issue. What do you think?

[the remark this answers is not in the document]
Comment (Tyratox) [reply fe1-web/src/core/components/Icon.tsx:38]: That would also be an option yes. This would almost result in the same interface `expo/vector-icons` provides, just with a restricted set of names. I am not sure whether this is better but if you think this is better I am happy to change it :)

File: fe1-web/src/features/wallet/screens/WalletCreateSeed.tsx (hunk 1 above)
Comment (Tyratox) [reply fe1-web/src/features/wallet/screens/WalletCreateSeed.tsx:87]: This works as well but `prettier` auto formats it to this `<Text> </Text>` 😅

Comment (Tyratox) [reply fe1-web/src/features/wallet/screens/WalletCreateSeed.tsx:87]: I guess I could add a `<Space />` component?

File: fe1-web/src/core/components/Icon.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (Xelowak) [reply fe1-web/src/core/components/Icon.tsx:38]: Let's go with your solution for now, we can still modify it later if needed.

**Review by Xelowak — APPROVED** [review]
LGTM! But why do tests fail? I'd like to merge, but I don't understand why it fails in other systems.

File: fe1-web/src/features/wallet/screens/WalletCreateSeed.tsx (hunk 1 above)
Comment (Xelowak) [reply fe1-web/src/features/wallet/screens/WalletCreateSeed.tsx:87]: Oh ok I didn't know that. If `prettier` thinks that it's fine, I guess we can trust it :)
